web/modules/custom/dpl_related_content/dpl_related_content.module

```diff
@@ -0,0 +1,292 @@
+    $tags_field_name = 'event_tags';
+    $categories_field_name = 'event_categories';
+    $branches_field_name = 'branch';
+  }
+
+  $tags = _dpl_related_content_get_term_ids($entity, $tags_field_name);
+  $categories = _dpl_related_content_get_term_ids($entity, $categories_field_name);
+  $branches = _dpl_related_content_get_term_ids($entity, $branches_field_name);
+  $event_ids = [];
+  $article_ids = [];
+
+  if (!empty($tags)) {
+    // First, let's look up related content, based only on tags.
+    $event_ids = _dpl_related_content_get_eventinstance_ids($current_uuid, $tags);
+    $article_ids = _dpl_related_content_get_article_ids($current_uuid, $tags);
+  }
+
+  // If we found less than 4 results, we'll add categories to the mix in
+  // addition to tags.
+  if ((count($event_ids) + count($article_ids) < 4) && !empty($categories)) {
+
+    $event_ids = _dpl_related_content_get_eventinstance_ids($current_uuid, $tags, $categories);
+    $article_ids = _dpl_related_content_get_article_ids($current_uuid, $tags, $categories);
+  }
+
+  // If the count is still under 4, and we're on an event-page, we'll look up
+  // events with the same location.
+  if ((count($event_ids) + count($article_ids) < 4) && ($entity instanceof EventInstance) && !empty($branches)) {
+
+    $event_ids = _dpl_related_content_get_eventinstance_ids($current_uuid, $tags, $categories, $branches);
+  }
+
+  // If the count is still under 4, we'll find the upcoming events, and the
+  // latest articles instead.
+  if (count($event_ids) + count($article_ids) < 4) {
+    $event_ids = _dpl_related_content_get_eventinstance_ids($current_uuid);
+    $article_ids = _dpl_related_content_get_article_ids($current_uuid);
+  }
+
+  // If we still have less than 4, we just won't display anything.
+  if (count($event_ids) + count($article_ids) < 4) {
+    return [];
+  }
+
+  $events = EventInstance::loadMultiple($event_ids);
+  $articles = Node::loadMultiple($article_ids);
+
+  // We want to combine events and articles into a single array, however,
+  // we want to interlace them, so it becomes mixed, rather than just
+  // all the articles and then all the events.
+  // We could also sort the merging based on the actual date values, but this
+  // will be too expensive for what it is worth.
+  $content = [];
+
+  // Remove array keys, so we can do the sorting.
+  $events = array_values($events);
+  $articles = array_values($articles);
+
+  $node_view_builder = \Drupal::entityTypeManager()->getViewBuilder('node');
+  $event_view_builder = \Drupal::entityTypeManager()->getViewBuilder('eventinstance');
+
+  // First, we find the longer array.
+  $length = max(count($events), count($articles));
+
+  for ($i = 0; $i < $length; $i++) {
+    if (isset($events[$i])) {
+      $content[] = $event_view_builder->view($events[$i], 'card');
+    }
+
+    if (isset($articles[$i])) {
+      $content[] = $node_view_builder->view($articles[$i], 'card');
+    }
+  }
+
+  return $content;
```

Comment:
Please consider the cache information here.

This does not seem to have any type of cache information available apart from what is provided by the individual renders. I would it to be dependent of a lot of different information - available articles, events, tags, the current time etc.

Perhaps it would make more sense to only cache this based on the time.